pkg/kubectl/rolling_updater.go

```diff
@@ -93,27 +79,433 @@ const (
+		if scaleUpFirst {
+			// Scale up/down. Initially this means scale up by the max increment.
+			scaledRc, err := r.scaleUp(newRc, oldRc, desired, increment, config)
+			if err != nil {
+				return err
+			}
+			newRc = scaledRc
+			scaleUpFirst = false
+		} // Otherwise, scale down/up.
+
+		// Scale down as much as possible while maintaining the minimum available
+		// amount. The scale-down informs us as to how much we can safely scale up
+		// next interval.
+		scaledRc, newIncrement, err := r.scaleDown(newRc, oldRc, desired, maxIncrement, minAvailable, availabilityCheck, config)
+		if err != nil {
+			return err
+		}
+		oldRc = scaledRc
+		increment = newIncrement
+
+		// Wait between down/up.
+		time.Sleep(config.UpdatePeriod)
+
+		// Scale up as much as possible.
+		scaledRc, err = r.scaleUp(newRc, oldRc, desired, increment, config)
+		if err != nil {
+			return err
+		}
+		newRc = scaledRc
+	}
+
+	// Housekeeping and cleanup policy execution.
+	return r.cleanup(oldRc, newRc, config)
+}
+
+// scaleUp scales up newRc to desired by increment. It will safely no-op as
+// necessary when it detects redundancy or other relevant conditions.
+func (r *RollingUpdater) scaleUp(newRc, oldRc *api.ReplicationController, desired, increment int, config *RollingUpdaterConfig) (*api.ReplicationController, error) {
+	// If we're already at the desired, do nothing.
+	if newRc.Spec.Replicas == desired {
+		return newRc, nil
+	}
+	// If the current safe increment is 0, we don't have capacity to scale up
+	// while maintaining in-place semantics, so do nothing.
+	if increment == 0 {
+		return newRc, nil
+	}
+	// If the old is already scaled down, go ahead and scale all the way up.
+	if oldRc.Spec.Replicas == 0 {
+		increment = desired - newRc.Spec.Replicas
+	}
+	// Increase the replica count, and deal with fenceposts.
+	newRc.Spec.Replicas += increment
+	if newRc.Spec.Replicas > desired {
+		newRc.Spec.Replicas = desired
+	}
+	// Perform the scale-up.
+	fmt.Fprintf(config.Out, "Scaling %s up to %d\n", newRc.Name, newRc.Spec.Replicas)
+	retryWait := &RetryParams{config.Interval, config.Timeout}
+	scaledRc, err := r.scaleAndWait(newRc, retryWait, retryWait)
+	if err != nil {
+		return nil, err
+	}
+	return scaledRc, nil
+}
+
+// scaleDown scales down oldRc to 0 by increment. It will safely no-op as
+// necessary when it detects redundancy or other relevant conditions.
+func (r *RollingUpdater) scaleDown(newRc, oldRc *api.ReplicationController, desired, maxIncrement, minAvailable int, checkType availabilityCheckType, config *RollingUpdaterConfig) (*api.ReplicationController, int, error) {
+	// Already scaled down; do nothing.
+	if oldRc.Spec.Replicas == 0 {
+		return oldRc, maxIncrement, nil
+	}
+	// Block until there are any pods available.
+	waitForAvailablePods := r.availabilityChecks[checkType]
+	oldAvailable, newAvailable, err := waitForAvailablePods(config.Interval, config.Timeout, oldRc, newRc)
```

Comment:
I wrestled with this a little, since depending on the interval settings the log output could get really spammy. If you think I'm being overly paranoid in that regard, I'll add the suggested output.